### PR TITLE
Improve separation between OrderManager and Connection.

### PR DESCRIPTION
**Pull request by pchote**
This splits the first two commits from #19632 into its own PR, as requested by @abcdefg30. See the original PR for more context on what these do and why.

### CODE REVIEW

File: OpenRA.Game/Network/OrderIO.cs [hunk 1]
```diff
@@ -9,70 +9,127 @@
...
+			}
+
+			frame = BitConverter.ToInt32(packet, 0);
+			syncHash = BitConverter.ToInt32(packet, 5);
+			defeatState = BitConverter.ToUInt64(packet, 9);
+			return true;
 		}
 
-		public static void Write(this BinaryWriter w, WPos pos)
+		public static bool TryParseOrderPacket(byte[] packet, World world, out int frame, out OrderPacket orders)
```
Comment (teinarss) [inline OpenRA.Game/Network/OrderIO.cs]: Parameter world is not used and can be removed.

File: OpenRA.Game/Network/OrderIO.cs [hunk 2]
```diff
@@ -9,70 +9,127 @@
  */
 #endregion
 
+using System;
 using System.Collections.Generic;
 using System.IO;
+using System.Linq;
```
Comment (teinarss) [inline OpenRA.Game/Network/OrderIO.cs]: Not needed!

File: OpenRA.Game/Network/OrderManager.cs [hunk 3]
```diff
@@ -125,74 +123,70 @@ void CacheTextNotification(TextNotification notification)
...
-				});
+			// HACK: The shellmap relies on ticking a disposed OM
+			if (disposed && World.Type != WorldType.Shellmap)
+				return;
+
+			pendingOrders.Remove(clientIndex);
 		}
 
-		Dictionary<int, byte[]> syncForFrame = new Dictionary<int, byte[]>();
+		Dictionary<int, (int SyncHash, ulong DefeatState)> syncForFrame = new Dictionary<int, (int, ulong)>();
```
Comment (teinarss) [inline OpenRA.Game/Network/OrderManager.cs]: can we move this up to the other fields in this pr or in a later?

**Comment by pchote** [comment]
Fixed.

File: OpenRA.Game/Network/ReplayRecorder.cs [hunk 4]
```diff
@@ -29,11 +29,10 @@ public sealed class ReplayRecorder
 
 		static bool IsGameStart(byte[] data)
 		{
-			if (data.Length > 4 && (data[4] == (byte)OrderType.Disconnect || data[4] == (byte)OrderType.SyncHash))
+			if (!OrderIO.TryParseOrderPacket(data, out _, out var orders))
 				return false;
 
-			var frame = BitConverter.ToInt32(data, 0);
-			return frame == 0 && data.ToOrderList(null).Any(o => o.OrderString == "StartGame");
+			return orders.GetOrders(null).Any(o => o.OrderString == "StartGame");
```
Comment (abcdefg30) [inline OpenRA.Game/Network/ReplayRecorder.cs]: Is the frame == 0 check not necessary? We could get the frame number from the first out parameter.

Comment (pchote) [reply OpenRA.Game/Network/ReplayRecorder.cs]: This was accidentally dropped during a rework. Fixed.

**Review by abcdefg30 — APPROVED** [review]

**Review by teinarss — APPROVED** [review]

**Comment by teinarss** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/acb358b636eb6615ddbf2ee22224e40d6dd2b680...577644f64f74368b5d21c111defdaea6681fdaaa

**Pull request merged**